Gh 232 Addition of Support for Multi Stream KCL2.x implementation.

siddharthjain210 opened the pull request:
Added support for Multi Stream Tracker in SI AWS. The older implementation has incorrected initialized the Stream Tracker Interface. This has been corrected. 

Added Adapter Test for Multi Stream Tests. Still needs some effort to enhance the tests, but due to the behavior of Enhanced Fan Out Registration, this takes too long. 

Need help on enhancing the tests. 

But this is working fine, i can see that the fix in the Stream Tracker has solved the problem and able to register the consumers in each stream.

pivotal-cla commented:
@siddharthjain210 Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-integration-aws&pullRequestId=234)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-integration-aws&pullRequestId=234) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-cla commented:
@siddharthjain210 Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-integration-aws&pullRequestId=234)!

artembilan reviewed (changes requested):
@siddharthjain210 ,

I also have just started to look into this.

So, here is what I understood so far:
1. We cannot do this initialization in  the ctor of `StreamsTracker` class: we just cannot perform Kinesis call so early on initialization phase. Therefore we have to deffer it directly to the `streamConfigList()` impl, or somehow cache. 
2. We have to use `describeStreamSummary()` API which is more efficient for our requirements.
3. The `streamCreationEpoch` can be extract from that `describeStreamSummary()` response as a `streamCreationTimestamp`. 
4. I'm not sure that we need extra test class to validate this situation: there should be just enough to enhance existing KCL test with one more stream. 
5. Not sure why we need logic for that least table. Why cannot we just rely on whatever KCL does for us by default?

Thank you for effort!

siddharthjain210 commented:
Thank you @artembilan for the review comments. 

Addressed review comments 1,2,3,5 Agree with the comments and have handled out. 

The purpose for adding the extra test class is to help understand how is it different from a single enhanced fan out consumer. 
In addition, we can only create one  `KclMessageDrivenChannelAdapter`.  
Also for test, we need 2 different stream and define the behavior. 
This makes it easier to understand both the behaviors, since its a little difficult.

I do need extra help to validate the test cases, due to the behavior of the FanOutRegistration, this makes it very hard to predict, when would the Consumers be available. Running in Debug buys some time and the tests run eventually.  Can you please take a look.

artembilan commented on src/main/java/org/springframework/integration/aws/inbound/kinesis/KclMessageDrivenChannelAdapter.java:
```diff
@@ -364,26 +365,34 @@ public Duration waitPeriodToDeleteFormerStreams() {
+			logger.info(LogMessage.format("Initializing stream config for Stream: [%s] for Multi Stream Tracker", streamName));
+
+			DescribeStreamSummaryResponse streamSummaryResponse = KclMessageDrivenChannelAdapter.this.kinesisClient.describeStreamSummary(DescribeStreamSummaryRequest.builder().streamName(streamName).build()).join();
+			return new StreamConfig(StreamIdentifier.multiStreamInstance(Arn.fromString(streamSummaryResponse.streamDescriptionSummary().streamARN()), streamSummaryResponse.streamDescriptionSummary().streamCreationTimestamp().toEpochMilli()),
```
`toEpochSeconds()`.
I heard somewhere that it AWS services deal in seconds for timestamps.

The `streamDescriptionSummary()` can be extracted to variable.

We also prefer the length of the line not more than 120 symbols.

artembilan commented on src/main/java/org/springframework/integration/aws/inbound/kinesis/KclMessageDrivenChannelAdapter.java:
```diff
@@ -364,26 +365,34 @@ public Duration waitPeriodToDeleteFormerStreams() {
+
 		@Override
 		public List<StreamConfig> streamConfigList() {
+			if (this.streamConfigs.size() == 0) {
```
I think if `== null` is better and then you can use `Stream` API to produce a list after mapping to `StreamConfig`

siddharthjain210 replied:
Yes, I agree, null check is better.

artembilan reviewed (changes requested):
It is good progress.

I'll look into your test when you address all other my concerns.

Thanks

siddharthjain210 commented on src/main/java/org/springframework/integration/aws/inbound/kinesis/KclMessageDrivenChannelAdapter.java:
```diff
@@ -260,7 +262,6 @@ protected void onInit() {
 						this.workerId,
 						this.recordProcessorFactory);
 	}
-
```
Been using Google Style using IntelliJ Shortcuts. This is sorted now.

siddharthjain210 commented on src/main/java/org/springframework/integration/aws/inbound/kinesis/KclMessageDrivenChannelAdapter.java:
```diff
@@ -353,7 +355,6 @@ public ShardRecordProcessor shardRecordProcessor(StreamIdentifier streamIdentifi
 	}
 
 	private final class StreamsTracker implements MultiStreamTracker {
-
```
this is done in all the 3 classes.

siddharthjain210 commented on src/main/java/org/springframework/integration/aws/inbound/kinesis/KclMessageDrivenChannelAdapter.java:
```diff
@@ -364,26 +365,34 @@ public Duration waitPeriodToDeleteFormerStreams() {
 		}
 
+		private StreamConfig buildStreamConfig(String streamName) {
+			logger.info(LogMessage.format("Initializing stream config for Stream: [%s] for Multi Stream Tracker", streamName));
```
removed.

artembilan replied:
I'm sorry, but I don't see changes in the PR.
Probably you have missed to push them...

artembilan commented on src/main/java/org/springframework/integration/aws/inbound/kinesis/KclMessageDrivenChannelAdapter.java:
```diff
@@ -364,26 +365,34 @@ public Duration waitPeriodToDeleteFormerStreams() {
+			if (this.streamConfigs.size() == 0) {
+				logger.info("Lazy loading the Stream Configs for Multi stream Tracker");
+				// Lazy loading the Stream Configs, only during inquiry.
+				Arrays.stream(KclMessageDrivenChannelAdapter.this.streams).forEach(streamName -> this.streamConfigs.add(buildStreamConfig(streamName)));
```
Looking into this one more time, I think we can leverage `reactor.core.publisher.Flux` API with all of its power of mapping, parallel processing and, of course, `cache()` in the end.
Let me know if you can do that or I'll address it on merge.

siddharthjain210 replied:
I Wont be able to do it immediately, would require a research and can take a good amount of time. Please feel free or we can take that as a separate PR.

artembilan replied:
No problem!
I'll push that as a separate commit and share with you here!

siddharthjain210 commented on src/main/java/org/springframework/integration/aws/inbound/kinesis/KclMessageDrivenChannelAdapter.java:
```diff
@@ -84,7 +87,6 @@
  * @author Herv? Fortin
  * @author Artem Bilan
  * @author Dirk Bonhomme
- *
```
This is resolved. Code pushed.

artembilan replied:
My review preference is like this:

if you agree with my change request, just make it and push.
I'll see it on the next review cycle.

If you are not agreed or have something to share, then comment: no need for reviews to have extra noise in their email box 😄

siddharthjain210 commented:
Please leave the review comments, I can take a look tomorrow and address the remaining.

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration-aws/commit/3a4f3552f551836f1b99d93241617943599c427c after some minor code clean up and tests improvements.

The KCL indeed has a long time for its initialization: doesn't look like we can affect it some how, but I fixed tests for different consumer groups, so they might not look into DynamoDB for old consumers from other test.

Thank you for contribution; looking forward for more!

artembilan commented:
So, here is the change I explained before based on Reactor `Flux`: https://github.com/spring-projects/spring-integration-aws/commit/2dd8a89f6a6687b9acedc1232f287741414a02eb

I also change one of our KCL tests to use a polling retrieval mode introduced in some previous commit.
